### PR TITLE
Add StremThru as streaming provider

### PR DESCRIPTION
**Pull request by MunifTanjim**
StremThru: https://github.com/MunifTanjim/stremthru

Example Config for StremThru:

```env
STREMTHRU_PROXY_AUTH=user-one:password-one,user-two:password-two                                                                 
STREMTHRU_STORE_AUTH="*:realdebrid:rd-api-key-one,user-two:realdebrid:rd-api-key-two"
```

Run by following: https://github.com/MunifTanjim/stremthru?tab=readme-ov-file#usage

## Release Notes

- **New Features**
	- Introduced a new streaming provider option, "StremThru," allowing users to configure additional settings including a Service URL.
	- Added fields for user credentials (email, password) and options for enabling watchlist catalogs.
	- Enhanced validation for the Service URL input to ensure proper formatting.

- **UI/UX Improvements**
	- Updated the configuration interface to include a new section for entering the Service URL, which is displayed based on user selection.
	- Adjusted CSS styles to accommodate new UI elements.

- **Bug Fixes**
	- Improved handling of provider-specific errors and validation for user input related to the new streaming service.

### CODE REVIEW

**Comment by mhdzumair** [comment]
Nice, lmk once you're done with the changes.

**Comment by mhdzumair** [comment]
is it ready for review? Please add validate credential function as well.

**Comment by MunifTanjim** [comment]
Not ready yet. I'm on vacation, will be back on 16th Nov and finish it after that.

**Comment by mhdzumair** [comment]
No worries, take your time & enjoy the vacation

File: streaming_providers/mapper.py [hunk 1]
```diff
@@ -61,6 +61,12 @@
     get_video_url_from_torbox,
     validate_torbox_credentials,
 )
+from streaming_providers.stremthru.utils import (
+    update_st_cache_status,
+    fetch_downloaded_info_hashes_from_st,
+    delete_all_torrents_from_st,
+    get_video_url_from_stremthru,
+)
```
[the remark this answers is not in the document]
Comment (mhdzumair) [reply streaming_providers/mapper.py:69]: Address this

File: db/schemas.py [hunk 2]
```diff
@@ -137,7 +137,9 @@ class StreamingProvider(BaseModel):
         "torbox",
         "premiumize",
         "qbittorrent",
+        "stremthru",
     ] = Field(alias="sv")
+    url: str | None = Field(default=None, alias="u")
     token: str | None = Field(default=None, alias="tk")
     email: str | None = Field(default=None, alias="em")
     password: str | None = Field(default=None, alias="pw")
```
[the remark this answers is not in the document]
Comment (mhdzumair) [reply db/schemas.py:145]: may be we dont need this

File: streaming_providers/stremthru/client.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+            "/v0/store/link/generate",
+            data={"link": link},
+            is_expected_to_fail=True,
+        )
+        if response.get("data", None):
+            return response["data"]
+        raise ProviderException(
+            f"Failed to create download link from StremThru {response["error"]}",
+            "transfer_error.mp4",
+        )
```
[the remark this answers is not in the document]
Comment (mhdzumair) [reply streaming_providers/stremthru/client.py:111]: Fix this. Double quote is overlapping.

**Comment by mhdzumair** [comment]
Address the above suggestion which contain :+1:

File: streaming_providers/stremthru/utils.py [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            for stream in streams:
+                stream.cached = any(
+                    torrent["status"] == "cached"
+                    for torrent in instant_availability_data["items"]
+                    if torrent.get("hash") == stream.id
+                )
+
+    except ProviderException:
+        pass
+
```
[the remark this answers is not in the document]
Comment (MunifTanjim) [reply streaming_providers/stremthru/utils.py:93]: @mhdzumair this error is ignored for all other providers 🤔

[the remark this answers is not in the document]
Comment (mhdzumair) [reply streaming_providers/stremthru/utils.py:93]: Sure you can ignore or keep its fine.

**Pull request merged**